design/aws-ecs-autoscaling-queue-worker.md

```diff
@@ -0,0 +1,216 @@
+
+To address issue [#2396](https://github.com/awslabs/aws-cdk/issues/2396), the AWS ECS CDK construct library should provide a way for customers to create a Queue Worker service (an AWS ECS/Fargate Service that processes items from an sqs queue). This construct will create an SQS Queue and enable autoscaling for the service based on cpu usage and the sqs approximateNumberOfMessagesVisible metric. This would mean adding new ECS CDK constructs `Ec2QueueWorkerService` and `FargateQueryWorkerService`, that would take in the necessary properties required to create a Task Definition, an SQS Queue as well as an ECS/Fargate Service with Auto Scaling rules.
+
+## General approach
+
+The new [`ecs.QueueWorkerServiceBase`], [`ecs.Ec2QueueWorkerService`] and [`ecs.FargateQueueWorkerService`] classes will include L3 constructs for:
+
+* Ec2QueueWorkerService
+* FargateQueueWorkerService
+
+A `QueueWorkerService` will create a task definition with the specified container (on both EC2 and Fargate). An `SQSQueue` will be created and autoscaling of the ECS Service will be dependent on both CPU as well as the SQS queue's `ApproximateNumberOfMessagesVisible` metric. 
+
+## Code changes
+
+Given the above, we should make the following changes to support queue workers on ECS (for both EC2 and Fargate):
+1. Create `QueueWorkerServiceBaseProps` interface and `QueueWorkerServiceBase` construct
+2. Create `Ec2QueueWorkerServiceProps` interface  and `Ec2QueueWorkerService` construct
+3. Create `FargateQueueWorkerServiceProps` interface  and `FargateQueueWorkerService` construct
+
+# Part 1: Create `QueueWorkerServiceBaseProps` interface  and `QueueWorkerServiceBase` construct  
+
+The `QueueWorkerServiceBaseProps` interface will contain common properties used to construct both the Ec2QueueWorkerService and the FargateQueueWorkerService:
+
+```ts
+/**
+ * Properties to define an Query Worker service
+ */
+export interface QueueWorkerServiceBaseProps {
+  /**
+   * Cluster where service will be deployed
+   */
+  readonly cluster: ICluster;
+
+  /**
+   * The image to start.
+   */
+  readonly image: ContainerImage;
+
+  /**
+   * The CMD value to pass to the container. A string with commands delimited by commas.
+   *
+   * @default none
+   */
+  readonly command?: string;
+
+  /**
+   * Number of desired copies of running tasks
+   *
+   * @default 1
+   */
+  readonly desiredTaskCount?: number;
+
+  /**
+   * The environment variables to pass to the container.
+   *
+   * @default true
+   */
+  readonly enableLogging?: boolean;
+
+  /**
+   * The environment variables to pass to the container.
+   *
+   * @default none
+   */
+  readonly environment?: { [key: string]: string };
+
+  /**
+   * A name for the queue.
+   *
+   * If specified and this is a FIFO queue, must end in the string '.fifo'.
+   *
+   * @default CloudFormation-generated name
+   */
+  readonly queueName?: string;
+
+  /**
+   * Maximum capacity to scale to.
+   *
+   * @default 2
```

Comment:
Default to desiredTaskCount?  If I set desiredTaskCount to 3, but don't specify maxScalingCapacity, this is awkward

Reply:
Wouldn't that mean by default, autoscaling is disabled? Is that the behaviour that we want?

Reply:
I think so, but I'm open to change

Reply:
I think making the default maxScalingCapacity = 2 * desiredTaskCount seems reasonable and if customers think that's too high, they can always pass it in. Thoughts?

Reply:
Sounds good to me

Reply:
Updated!